ssl/ech/ech_internal.c

```diff
@@ -160,4 +214,1260 @@ int ossl_ech_conn_init(SSL_CONNECTION *s, SSL_CTX *ctx,
+     * if we're a client we'll fallback to hash from the chosen ciphersuite
+     * that means ECH acceptance depends on no bidding down, but that's ok
+     */
+    rv = ossl_ech_get_sh_offsets(shbuf, shlen, &extoffset, &echoffset,
+                                 &echtype);
+    if (rv != 1 || extoffset < 3) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, SSL_R_ECH_REQUIRED);
+        return 0;
+    }
+    cipheroffset = extoffset - 3;
+    cipherchars = &shbuf[cipheroffset];
+    c = ssl_get_cipher_by_char(s, cipherchars, 0);
+    if (c == NULL /* fuzzer fix */
+        || (md = (EVP_MD *)ssl_md(s->ssl.ctx, c->algorithm2)) == NULL) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    *rmd = md;
+    return 1;
+}
+
+/*
+ * make up a buffer to use to reset transcript
+ * for_hrr is 1 if we've just seen HRR, 0 otherwise
+ * shbuf is the output buffer
+ * shlen is the length of that buffer
+ * tbuf is the output buffer
+ * tlen is the length of that buffer
+ * chend returns the offset of the end of the last CH in the buffer
+ * fixedshbuf_len returns the fixed up length of the SH
+ * return 1 for good, 0 otherwise
+ */
+int ossl_ech_make_transcript_buffer(SSL_CONNECTION *s, int for_hrr,
+                                    const unsigned char *shbuf, size_t shlen,
+                                    unsigned char **tbuf, size_t *tlen,
+                                    size_t *chend, size_t *fixedshbuf_len)
+{
+    unsigned char *fixedshbuf = NULL, *hashin = NULL, hashval[EVP_MAX_MD_SIZE];
+    unsigned int hashlen = 0, hashin_len = 0;
+    EVP_MD_CTX *ctx = NULL;
+    EVP_MD *md = NULL;
+    WPACKET tpkt, shpkt;
+    BUF_MEM *tpkt_mem = NULL, *shpkt_mem = NULL;
+
+    /*
+     * SH preamble has bad length at this point on server
+     * and is missing on client so we'll fix
+     */
+    if ((shpkt_mem = BUF_MEM_new()) == NULL
+        || !WPACKET_init(&shpkt, shpkt_mem)) {
+        BUF_MEM_free(shpkt_mem);
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    if (!WPACKET_put_bytes_u8(&shpkt, SSL3_MT_SERVER_HELLO)
+        || (s->server == 1 && !WPACKET_put_bytes_u24(&shpkt, shlen - 4))
+        || (s->server == 1 && !WPACKET_memcpy(&shpkt, shbuf + 4, shlen -4))
+        || (s->server == 0 && !WPACKET_put_bytes_u24(&shpkt, shlen))
+        || (s->server == 0 && !WPACKET_memcpy(&shpkt, shbuf, shlen))
+        || !WPACKET_get_length(&shpkt, fixedshbuf_len)) {
+        BUF_MEM_free(shpkt_mem);
+        WPACKET_cleanup(&shpkt);
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    fixedshbuf = OPENSSL_malloc(*fixedshbuf_len);
+    if (fixedshbuf == NULL) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        BUF_MEM_free(shpkt_mem);
+        WPACKET_cleanup(&shpkt);
+        goto err;
+    }
+    memcpy(fixedshbuf, WPACKET_get_curr(&shpkt) - *fixedshbuf_len,
+           *fixedshbuf_len);
```

Comment:
Not sure I see why this separate buffer with memcpy is necessary. Why not just use the buffer from the `BUF_MEM` direct?

Reply:
In the event of HRR that buffer is stashed for use when the SH is rx'd but yeah I guess there's no need for the malloc in the overwhelmingly non-HRR cases so changed it about